### PR TITLE
JetBrains: Cody: Switch to master-layout-based settings UI

### LINKED ISSUE #56381: JetBrains: switch to master-layout-based settings UI
Following up from https://sourcegraph.slack.com/archives/C059N5FRYG3/p1692104355148129

Currently, the plugin is using "Group header" based design. We should switch to a Tab-based design instead to make it clearer that Cody and Code Search features are separate. I've seen multiple users expecting that the code-search related settings (remote URL) are used to configure how Cody infers embeddings URL.

Here is an example how the Scala plugin uses tabs to group sections of settings

Image

The Sourcegraph plugin will need three tabs: Authentication, Cody and Code Search

Image

### PR DESCRIPTION
**Pull request by Gedochao**
Fixes #56381 

https://github.com/sourcegraph/sourcegraph/assets/18601388/12ab0358-c6bc-4406-90c6-5fde3aa14575

## Test plan
- ensure all the settings work (I am aware  that's a lot of edge scenarios, yeah)

### CODE REVIEW

**Review by vdavid — APPROVED** [review]
- I like the UI
- I like how the rewritten UIs look in Kotlin—much cleaner than the Java version was
- Breaking up the change listeners looks great
- Looks like storage didn't change at all, which makes this change quite safe

File: client/jetbrains/src/main/kotlin/com/sourcegraph/cody/config/notification/ChangeListener.kt [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import com.intellij.util.messages.MessageBusConnection
+import com.sourcegraph.find.browser.JavaToJSBridge
+
+abstract class ChangeListener(protected val project: Project) : Disposable {
+    protected var connection: MessageBusConnection? = null
+    var javaToJSBridge: JavaToJSBridge? = null
+    protected val logger = Logger.getInstance(ChangeListener::class.java)
+    private val bus: MessageBus = project.messageBus
+
+    init {
```
Comment (szymonprz) [inline client/jetbrains/src/main/kotlin/com/sourcegraph/cody/config/notification/ChangeListener.kt:16]: you can set the connection directly in the field and the type of the MessageBusConnection can be not null

**Review by szymonprz — APPROVED** [review]

File: client/jetbrains/src/main/kotlin/com/sourcegraph/cody/config/notification/AccountSettingChangeListener.kt [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+                    } else {
+                        // Stopping the agent is idempotent, so it's OK if we call stopAgent multiple times.
+                        CodyAgentProjectListener.stopAgent(project)
+                    }
+
+                    // Notify Cody Agent about config changes.
+                    val agentServer = CodyAgent.getServer(project)
+                    if (ConfigUtil.isCodyEnabled() && agentServer != null) {
+                        agentServer.configurationDidChange(ConfigUtil.getAgentConfiguration(project))
+                    }
```
[the remark this answers is not in the document]
Comment (Gedochao) [reply client/jetbrains/src/main/kotlin/com/sourcegraph/cody/config/notification/AccountSettingChangeListener.kt]: I guess it shouldn't be here, removed it

File: client/jetbrains/src/main/kotlin/com/sourcegraph/cody/config/notification/ChangeListener.kt (hunk 1 above)
Comment (Gedochao) [reply client/jetbrains/src/main/kotlin/com/sourcegraph/cody/config/notification/ChangeListener.kt]: fixed

**Pull request merged**